cpp/src/request_tag_multi.cpp

```diff
@@ -127,8 +114,10 @@ void RequestTagMulti::recvFrames()
     for (size_t i = 0; i < h.nframes; ++i) {
       auto bufferRequest = std::make_shared<BufferRequest>();
       _bufferRequests.push_back(bufferRequest);
-      const auto bufferType  = h.isCUDA[i] ? ucxx::BufferType::RMM : ucxx::BufferType::Host;
-      auto buf               = allocateBuffer(bufferType, h.size[i]);
+      const auto bufferType = h.isCUDA[i] ? ucxx::BufferType::RMM : ucxx::BufferType::Host;
+      auto buf              = allocateBuffer(bufferType, h.size[i]);
+      bufferRequest->requestTagMulti =
+        std::static_pointer_cast<RequestTagMulti>(shared_from_this());
```

Comment:
Bit confused here. `_bufferRequests` is a member variable of `this` (a `RequestTagMulti`), and this seems to be making a reference cycle by saving a shared pointer to `this` in the buffer request that is then saved on `this`.

So why does this not create a situation like:
```
this---->bufferRequest
 ^            |
 |            |
 `------------'
```
?

Reply:
Is it that someone else then gets hold of the `bufferRequest` objects (rather than holding on to the `RequestTagMulti` object), so something like this might happen (pseudo-code):

```
multi_request = ...;
multi_request->recvFrames(...);
buf_requests = multi_request->get_requests(...);
// lose reference to multi_request
...;
buf_requests->complete(); // uh-oh, multi_request which is needed might be gone
```

Reply:
IOW, whose job is it to hold on to the `multi_request`, and how does this change help, except by leaking?

Is this situation now not like:
```c++
#include <memory>
#include <iostream>
struct B;

struct A {
  std::shared_ptr<B> b;
  A() { std::cout << "Created A" << std::endl; }
  ~A() { std::cout << "Deleted A" << std::endl; }
};

struct B {
  std::shared_ptr<A> a;
  B() { std::cout << "Created B" << std::endl; }
  ~B() { std::cout << "Deleted B" << std::endl; }
};

int main(void)
{
  auto a = std::make_shared<A>();
  auto b = std::make_shared<B>();
  b->a = a;
  a->b = b;
}
```
Which never calls the destructors for `A` and `B`?

Reply:
Thanks @wence- for realizing this and taking the time to discuss that with me last week, you were spot on in your review. I've since reverted those changes and instead ensure that the callback takes a `shared_from_this()` reference [here](https://github.com/rapidsai/ucxx/pull/62/files#diff-525902688d0bc5f703bd2edfe719a1fe3593546160de89ab66284271c596bc48R105-R111) and [here](https://github.com/rapidsai/ucxx/pull/62/files#diff-b15e3d4a4461df6e15af9a85397fb8f5d2340cc6062d532405e8de68074855d8R150-R157), and thus ensure the object remains valid at least until `ucxx::Request::callback()`/`ucxx::RequestTagMulti::markCompleted()` finish running, independent of the user releasing that as soon as the status gets set. This should also prevent any cycles.